### DISCUSSION TITLE
Missing entity type information. sObject requires a separate 'type' field be sent.

### DISCUSSION THREAD

**Discussion by Rana-xD** [jsforce/jsforce#1478] [Q&A]
Hello
I tried to use soap.create() function. But, I got the error. "Missing entity type information. sObject requires a separate 'type' field be sent." May I know the format for sObject?

Here is my code.

`const jsforce = require('jsforce');`
`const fs = require('fs');`

 ` let username = "usernamet";`
  `let password = "password";`
  
 ` const conn = new jsforce.Connection({});`
 ` let sObject = JSON.parse(fs.readFileSync('data.json'));`
`  conn.login(username, password, function(err, userInfo) {`
          ` if (err) { return console.error(err); }`
          `conn.soap.create(sObject,function(err,res){`
                ` if (err) { return console.error(err); }`
                     `console.log(res);`
                ` })`
           ` // ...`
  `});`

**Comment by Ahd-3hd** [comment]
facing the same issue with merge api:

```
export const handler = async (event) => {
  const accountObject = await conn
    .sobject('Account')
    .findOne({ Id: 'xxxxxxxxxxxx' });

  const result = await conn.soap.merge({
    masterRecord: accountObject,
    recordToMergeIds: ['xxxxxxxxxxxx'],
  });

  console.log(result);
};

```

```
offline: Failure: Missing entity type information. sObject requires a separate 'type' field be sent.
soapenv:Client: Missing entity type information. sObject requires a separate 'type' field be sent.
    at module.exports.HttpApi.getError (C:\Users\Ahd\Desktop\sfdc_integration\node_modules\jsforce\lib\http-api.js:250:13)
    at C:\Users\Ahd\Desktop\sfdc_integration\node_modules\jsforce\lib\http-api.js:95:22
    at tryCallOne (C:\Users\Ahd\Desktop\sfdc_integration\node_modules\promise\lib\core.js:37:12)
    at C:\Users\Ahd\Desktop\sfdc_integration\node_modules\promise\lib\core.js:123:15
    at flush (C:\Users\Ahd\Desktop\sfdc_integration\node_modules\asap\raw.js:50:29)
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
```

**Comment by mahnoor9** [comment]
facing the same issue

**Comment by Ahd-3hd** [comment]
a work around would be to use the REST api to call a custom apex class that merges the wanted records.

**Comment by mahnoor9** [comment]
i am pretty new to this, can you share a code example of this solution?

**Comment by mahnoor9** [comment]
i am doing something like this to create a record inside the salesforce platform using soap API: 
`conn.soap.create(obj , function(err,result){
      if(err){console.log(err)}
      console.log(result);
    })`

**Comment by Ahd-3hd** [comment]
> i am doing something like this to create a record inside the salesforce platform using soap API: `conn.soap.create(obj , function(err,result){ if(err){console.log(err)} console.log(result); })`

I lost the code to be honest, but what you can do is utilize APEX rest api.
in salesforce you can create your own api end point as a class and give it a decorator and methods like a normal rest api.
this is an apex code example to create a rest api end point to get and update duplicated records:
you need to find how to merge records using `APEX` since merging will happen via apex code.

```
//handle ids of same type
@RestResource(urlMapping='/findDuplicatesByIds')
global with sharing class SFDCDuplicatesRecords {
  
    @HttpPost
    global static String doPost() {
        RestRequest req = RestContext.request;
        RestResponse res = RestContext.response;
        String payload = RestContext.request.requestBody.toString();
		List<ID> idList = (List<ID>) JSON.deserialize(payload, List<ID>.class);
        return JSON.serialize(Datacloud.FindDuplicatesByIds.findDuplicatesByIds(idList)[0].duplicateResults[0].matchResults[0].matchRecords);
    }
}

// handle multiple ids with different types
// refactor needed
@RestResource(urlMapping='/sfdc-find-duplicate-by-ids')
global with sharing class SFDCDuplicatesRecords {
  
    @HttpPost
    global static String doPost() {
        RestRequest req = RestContext.request;
        RestResponse res = RestContext.response;
        String payload = RestContext.request.requestBody.toString();
        
        List<ID> idList = (List<ID>) JSON.deserialize(payload, List<ID>.class);
        List<Object> result = new List<Object>();
        for (ID id : idList){
            List<ID> ids = new List<ID>();
            ids.add(id);
            result.add(Datacloud.FindDuplicatesByIds.findDuplicatesByIds(ids));
        }
        return JSON.serialize(result);
    }
}
```

https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotations_rest.htm

and then you can make a fetch frequest on that endpoint using either axios or jsforce.

**Comment by mahnoor9** [comment]
Thanks a lot, the issue is resolved!

**Comment by dubreuil534** [comment]
@Ahd-3hd 
I found the solution to use this.connection.soap.merge(). 

- In the masterRecord object, we need to specify the type of object and the id of the master Record.

See exemple: 

  const mergeRequest: MergeRequest = {
            masterRecord: { type: 'Account', Id: '001Aq000005YOI4IAO' },
            recordToMergeIds: ['001Aq000005GQ3iIAG'],
        };
        await this.connection.soap.merge(mergeRequest);

if the type object is not insert. then this error message is throw : Missing entity type information. sObject requires a separate 'type' field be sent

**Comment by EddyPermana22** [comment]
> Thanks a lot, the issue is resolved!

Hi @mahnoor9, can you share the solution?
I facing the same problem and need to insert records using SOAP
